File: sdk-api/src/test/java/dev/restate/sdk/SleepTest.java
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk;

import static dev.restate.sdk.JavaBlockingTests.testDefinitionForService;

import dev.restate.sdk.common.Serde;
import dev.restate.sdk.core.SleepTestSuite;
import dev.restate.sdk.core.TestDefinitions.TestInvocationBuilder;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;

public class SleepTest extends SleepTestSuite {

  protected TestInvocationBuilder sleepGreeter() {
    return testDefinitionForService(
        "SleepGreeter",
        Serde.VOID,
        JsonSerdes.STRING,
        (ctx, unused) -> {
          ctx.sleep(Duration.ofSeconds(1));
          return "Hello";
        });
  }

  protected TestInvocationBuilder manySleeps() {
    return testDefinitionForService(
        "ManySleeps",
        Serde.VOID,
        Serde.VOID,
        (ctx, unused) -> {
          List<Awaitable<?>> collectedAwaitables = new ArrayList<>();

          for (int i = 0; i < 10; i++) {
            collectedAwaitables.add(ctx.timer(Duration.ofSeconds(1)));
          }

          Awaitable.all(
                  collectedAwaitables.get(0),
                  collectedAwaitables.get(1),
                  collectedAwaitables
                      .subList(2, collectedAwaitables.size())
                      .toArray(Awaitable[]::new))
              .await();

          return null;
        });
  }
}

File: sdk-api-kotlin/src/test/kotlin/dev/restate/sdk/kotlin/SleepTest.kt
// Copyright (c) 2023 - Restate Software, Inc., Restate GmbH
//
// This file is part of the Restate Java SDK,
// which is released under the MIT license.
//
// You can find a copy of the license in file LICENSE in the root
// directory of this repository or package, or at
// https://github.com/restatedev/sdk-java/blob/main/LICENSE
package dev.restate.sdk.kotlin

import dev.restate.sdk.core.SleepTestSuite
import dev.restate.sdk.core.testservices.GreeterGrpcKt
import dev.restate.sdk.core.testservices.GreetingRequest
import dev.restate.sdk.core.testservices.GreetingResponse
import dev.restate.sdk.core.testservices.greetingResponse
import io.grpc.BindableService
import kotlin.time.Duration.Companion.milliseconds
import kotlinx.coroutines.Dispatchers

class SleepTest : SleepTestSuite() {
  private class SleepGreeter :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateKtComponent {
    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val ctx = ObjectContext.current()
      ctx.sleep(1000.milliseconds)
      return greetingResponse { message = "Hello" }
    }
  }

  override fun sleepGreeter(): BindableService {
    return SleepGreeter()
  }

  private class ManySleeps :
      GreeterGrpcKt.GreeterCoroutineImplBase(Dispatchers.Unconfined), RestateKtComponent {
    override suspend fun greet(request: GreetingRequest): GreetingResponse {
      val ctx = ObjectContext.current()
      val awaitables = mutableListOf<Awaitable<Unit>>()
      for (i in 0..9) {
        awaitables.add(ctx.timer(1000.milliseconds))
      }
      awaitables.awaitAll()
      return greetingResponse {}
    }
  }

  override fun manySleeps(): BindableService {
    return ManySleeps()
  }
}
